design-doc.md

```diff
@@ -0,0 +1,128 @@
+
+The default design stores binary-serialized entires in append-only files. The framework keeps one *active*, currently appended to, file, and a directory of already *finalized* files.
+
+Default file finalization policy uses file size cap, max. number of entries per one file and max. age of a file. Default settings are 100M entries, as 10MB files, containing no more than 24 hours time window of a file.
+
+For faster lookup, finalized files include first and last indexes, as well as first and last order keys, as part of the filename. Thus, for finalized files, only their names have to be scanned in order to be able to efficiently create listeners.
+
+
+## Dataflow
+
+Fundamentally, a stream supports two operations:
+
+1. **Listen**: Run certain handler for each entry from the stream, potentially indefinitely.
+2. **Publish**: Add a new entry to the stream.
+
+### Listeners
+
+Listener is a callback that is invoked per each stream entry, in the order the entries have been added.
+
+By default, the listener starts at the beginning of a stream and never stops.
+
+As for the starting enrty for the listener, there are ways to start the listener from a specific point in the past, specifically, from an entry with certain index, or from an entry with certain order key. The former is used for stream replication, the latter is used for TailProduce jobs, where the producer guarantees that its state at certain "time == order key" is agnostic with respect to how many entries preceeding this order key minus a fixed, specified time window width.
+
+As for the end of the listener, it is possible to request the listener to terminate automatically after certain number of entries have been scanned or after certain order key has been reached. On top of that, the listener itself can choose to stop itself when certain criteria has been met. Since listeners are independent, the termination criteria does not have to be consistent, even within the very same listener.
+
+Implementation-wise, listening to a stream spawns a dedicated thread. (`TODO(dkorolev): The design enforces the users to be "good citizens", and not spawn listeners in large quantities, however, in certain cases one thread per listener might not be the most efficient approach. I should document it in a bit more detail.`).
+
+Being run in dedicated threads, listeners never interfere with each other and with the publisher. It is legit for a very slow listener to run in parallel with the one that is expected to always be caught up with the most recently added entires.
+
+A per-entry handler of a listener has the means to tell how far "behind" on a stream it is, both index-wise and order-key-wise. This is important for serving guarantees: for example, when a listener is responsible for serving external requests, the common design is to notify its master or load balancer that it is healthy as it is caught up and that it still is in the "replay log" mode.
+
+### Publishing
+
```

Comment:
In the previous section there is a clear starting point: "Listener is ...". It would be great to have the same structure here, consistent with the the statement from the **Terminology** paragraph, declaring the concepts:
`Entries can be appended to streams using **publishers** and scanned with **listeners**.`